### ISSUE TITLE
stability: Rebalances must be atomic

### ISSUE THREAD

**Issue by bdarnell** [cockroachdb/cockroach#12768]
Consider a deployment with three datacenters (and a replication configuration that places one replica of each range in each DC). During a rebalance, a range that normally requires a quorum of 2/3 will briefly change to a configuration with a 3/4 quorum and two replicas in one DC. If that DC (or the network) fails before the fourth replica is removed, the range becomes unavailable (If rebalancing downreplicated before upreplicating, then the intermediate state would have a quorum of 2/2 and a failure of *either* datacenter would lead to unavailability and much harder recovery). While the window is fairly small, there is a risk that an ill-timed DC failure could make one or more ranges unavailable for the duration.

The most general solution to this problem is to implement "joint consensus", the original raft membership change proposal (it is described in the original raft paper and in section 4.3 of Diego's dissertation). Etcd implements a much simpler membership change protocol which has the limitation that only one change can be made at a time (leading to the brief exposure of the intermediate 3/4 state). Joint consensus would let us make these rebalancing changes atomically. 

We might also be able to mitigate the problem by using ideas from [Flexible Paxos](https://fpaxos.github.io/). Flexible Paxos shows that committing entries can be done with a bare majority of nodes, so as long as the leader is not in the failed datacenter the removal of the fourth node can be completed while the DC is down and the range is restored to its 2/3 state. However, if the leader were in the failed DC then the two surviving ones would be unable to make progress since they would have to assume that the former leader is still making progress on the other side of its partition. I'm not sure if there's a full solution here or not.

Running with a higher replication factor (5 replicas in 3 DCs) could also mitigate the problem if the rebalancer were aware of it (so when the range goes from 3/5 to 4/6, those six replicas are guaranteed to be two per DC). This might be a quick fix to prevent this problem from striking a critical system range. Running with more DCs than replicas (3 replicas in 4 DCs, so you never need to place two replicas in the same DC) also avoids this problem without increasing storage costs, but it has the significant downside that all rebalancing must happen over the more expensive inter-DC links.

**Comment by arjunravinarayan** [comment]
If we're using flexible paxos, we can make it work as long as we first ensure that the leader is *not* on the datacenter where the rebalance is taking place. So if we denote datacenters by numbers and replicas by letters, if we are going from {1A, 2B, 3C} to {1A, 2B, 3D}, we just need to ensure that A or B is the leader before upreplicating to {1A, 2B, 3C, 3D}. Now if we have a datacenter failure at 1 or 2, we have enough replicas (three) to elect a leader. If we have a datacenter failure at 3, we still have our leader so we can upreplicate again to a 5-replica range.

**Comment by bdarnell** [comment]
Yeah, we'd need to ensure both that A or B is the leader and that C cannot become leader while the upreplication is in flight.

There are also performance considerations here (which we do not currently take into account): the new replica gets its snapshot from the leader, so for performance we'd want to have 3C be the leader when 3D is added. (or we introduce some way to send preemptive snapshots from a follower)

**Comment by bdarnell** [comment]
We have a real-world instance of this: a node crashed (in a persistent way so it crashes again each time it's restarted) while the range was only configured for two replicas. This required extreme efforts to revive the downed node instead of simply letting the cluster recover normally.

**Comment by a-robinson** [comment]
While it was only configured for two replicas? So this was when up-replicating from 1 to 3? I'm drawing a blank on when else a range would be configured to have two replicas during rebalancing (other than when removing a replica from a dead store and adding a new one to replace it, which this doesn't sound like).

**Comment by bdarnell** [comment]
I think this was while removing a replica from a dead store. So technically this was a two-simultaneous-failure event, but the first failure was transient while the second was persistent. If the replica change had been atomic, we'd have been able to recover when the first transient failure resolved.

**Comment by a-robinson** [comment]
Yeah, we need to fix that, although the fix for that will be different from / simpler than the fix for what this issue was originally describing. We really just need to up-replicate before removing the dead replica, not to make anything atomic.

For the record, we hit this issue or removing a dead replica before adding its replacement a couple months ago as well - https://github.com/cockroachdb/cockroach/issues/25392.

**Comment by tschottdorf** [comment]
I folded https://github.com/cockroachdb/cockroach/issues/2067 (can't change the membership with three replicas in three nodes cluster) into this issue.

**Comment by bdarnell** [comment]
While a complete fix requires the implementation of joint consensus, there may be things we can do to make partial improvements. There are a few factors that may be leaving ranges in this over-replicated, under-diversified state longer than necessary.

- In the allocator, removing an "extra" replica is given a low priority, but as this example shows the resulting under-diversification can be as severe a problem as a dead node.
- The replicate queue is a single thread that is blocked while sending preemptive snapshots, preventing other ranges from downreplicating for the duration of a snapshot. (we can easily add more concurrency in this queue, but as long and adds and removes go through the same pool we'll likely end up with all threads blocked in snapshot sending)
- After adding a replica, we just send the range back through the replicate queue to remove the extra one. That removal is not guaranteed to be the next thing we do. Even if it is, we only make one attempt, and if that fails the extra replica will persist until the next scanner cycle.

I think the last issue is probably most significant - a cluster that experiences one failure to remove an extra replica per ten minutes will usually have an overreplicated range and be vulnerable to region/AZ failure. There are a number of ways this can happen. For example, we try not to remove a node that is "behind", but I think the combination of heuristics here will err on the side of not removing anything (especially considering that the newly-added replica is likely not caught up yet, and if the queue is quick enough we reach this point while only 2/3 of the original replicas have acked the replica change itself).

**Comment by bdarnell** [comment]
![image](https://user-images.githubusercontent.com/160562/54880529-d3711d80-4e1b-11e9-8587-53f82a1b6d60.png)

Here's the new metric added in #36084 on a 7-node single-region cluster running kv0 for 5 minutes. At its peak, nearly a third of ranges were over-replicated, and this is slow to correct. The load stopped at 13:58, but this had no visible impact on the slope of the over-replication line. It looks like the slope did change, slowing down around 14:07 with 18 overreplicated ranges remaining. These are clearing very slowly.

This is consistent with the theory above: the replication queue is making a single attempt to downreplicate, and if that fails it's being left for the next scanner cycle. Most of the ranges became over-replicated quickly, then the number declined linearly over the next 10 minutes (one scanner cycle), followed by a shallower decline for any that were missed by that cycle.

The number of overreplicated ranges here is far higher than expected - we should just have a few of them flitting in and out of existence temporarily. If 1/3 of replicas are overreplicated, and 1/3 of those have their extra replica in an AZ that fails, this would result in 11% data unavailability due to an AZ failure (I have no idea how or whether these numbers extend to other scenarios). That's not good, but it does suggest that we have a lot of room to improve those numbers with improvements to the replication queue in advance of the full joint consensus solution.

**Comment by bdarnell** [comment]
https://github.com/cockroachdb/cockroach/blob/5a746073c3f8ede851f37dd895cf1a91d6dcc3cf/pkg/storage/replicate_queue.go#L373-L385

If I add a retry loop around this code, the overreplicated ranges almost entirely go away: 

![image](https://user-images.githubusercontent.com/160562/54882799-1ccd6700-4e34-11e9-892b-d515355b9972.png)

We have 3 overreplicated ranges at the start of the run (vs hundreds without the retry loop), and they're gone within the first minute of the test. This gives us empirical confirmation that most of the overreplicated ranges were coming from the range temporarily not having enough replicas that are known to be up to date.

**Comment by tbg** [comment]
Great find. Are you considering adding a max ~2*raft hartbeat interval poll here?
Are you considering upping the priority for removals? We'd have to make sure there's no scenario in which a number of removals can remain queued, not in purgatory, and yet failing.
Or we extend the queue to allow processing of 1 addition and N removals in parallel, but I suspect that the queue infra does not lend itself well to these things and would further get complicated by that kind of change.

**Comment by andreimatei** [comment]
> In the allocator, removing an "extra" replica is given a low priority, but as this example shows the resulting under-diversification can be as severe a problem as a dead node.

@bdarnell what did you mean by the comment above? How are we vulnerable to a single node death when we're in the 4 replicas state?

**Comment by tim-o** [comment]
Zendesk ticket #[3282](https://cockroachdb.zendesk.com/agent/tickets/3282) has been linked to this issue.

**Comment by tim-o** [comment]
Zendesk ticket #[3094](https://cockroachdb.zendesk.com/agent/tickets/3094) has been linked to this issue.

**Comment by tbg** [comment]
- [ ] flip atomic replication changes to default-on on master
- [ ] test recovery from snapshot
- [ ] test recovery from initial state
- [ ] teach replicate queue to emit a swap when one is appropriate
- [ ] stress test for DC outage during rebalancing in supposedly "safe" config

**Comment by tbg** [comment]
I'm hitting a branching point on which I want to make a decision soon. I think I've come to terms with which way I think things should move, but that should be bounced off a few folks here.

The two options are about how the joint state is encoded and managed:

1. track the joint config via the range descriptor, and transition out via an additional descriptor change txn
1. let raft auto-transition us out of the joint config

The latter is what I initially prototyped. We run the change replicas txn to completion, resulting in a descriptor that describes the final state, whereas the real configuration is joint. Replicas needs to be able to synthesize its ConfState from the persisted state at any point in time, so when entering the joint state we need to persist what is essentially the previous set of voters (i.e. the old descriptor), and clear it when leaving it again (i.e. when observing the raft command).
This was initially chosen because it seemed leaky to expose the joint state everywhere into the system, but I'm now skeptical that approach is right because it's complex to teach and maintain. For example, when in the joint state, the replicaGC queue will naively remove voters because... they're not in the range descriptor any more. Raft message routing needs special code because the translation from replicaID to recipient store is done via the descriptor. And lots of other code that cares about "voters" probably wants to know about these "hidden" voters too, making this a future recurrent source of surprises. These are already potentially enough points to convince me that this isn't the way to go, but it gets worse: splits and merges need to transactionally interact with the joint state; they can't do so through the descriptor but have to do it via that extra hidden "joint state". But that state enjoys no protection from latches since it's manipulated from commands that originate within raft. So we need to build awkward crutches around that. If we don't, a merge might start and commit before the joint config was exited, and the outgoing voters will be asked to merge a nonexistent right hand side.

The former option is more straightforward: the change replicas trigger txn writes a descriptor that encodes the joint config (i.e. there's an new VoterOutgoing replica descriptor type or something like that). In the common case, right after that we run an extra txn with a trigger that transitions out of the joint config (i.e. updates the descriptor one more time). This is all conceptually much more straightforward: there's a new replica type and everyone needs to handle it appropriately. The replicaGC queue will observe these replicas; we must make sure to not let outgoing voters request the lease, etc. Visiblity into the joint state is good because it's encoded on the descriptor, both on the range and in the meta ranges.
On what is perhaps the minus side, we've now picked up the burden of making sure we transition out of this state promptly even through crashes, etc., but this seems straightforward enough to achieve.

**Comment by tbg** [comment]
Ran @bdarnell's experiment again on my branch that has atomic rebalancing default to on:

<img width="737" alt="image" src="https://user-images.githubusercontent.com/5076964/64339270-ecddde80-cfe3-11e9-8bfb-a7a1f0fb290a.png">

There's little to no over-replication going on. What little bits and pieces we see is (I think) due to calls to RelocateRange:

https://github.com/cockroachdb/cockroach/blob/78e4c884a1dfac0e1b478c183c7013e9fd55763b/pkg/storage/replica_command.go#L1990-L1992

which has lots of complicated logic which for historical reasons does one-change-at-a-time. What I don't understand here is why this talks to the allocator at all; the targets are predetermined by the input. @danhhz this looks like you know this code somewhat, do you have an answer for me? The new code will still be nasty because we can't remove the leaseholder (https://github.com/cockroachdb/cockroach/issues/40333), but it can at least use atomic changes for individual add-remove pairs which in particular means that as long as we're not trying to change the cardinality of the replica set we'll never do so.

I'm not sure how good the over-replicated metric is in noticing smaller short-lived infractions. It's on a 10s interval, so unless an overreplication lasts for "seconds", it may very well be missed. Either way, I was looking at the logs while the rebalancing happened and I'm seeing all of it use joint consensus (outside of calls to AdminRelocateRange).

Curiously I observed the opposite problem: we had 17 under-replicated ranges that dropped to zero over ~10 minutes (one scanner cycle). I looked at one of them and it was r5 which was waiting go to from 3 to 5 replicas (it's a system range).

We'll obviously try to improve this as the release comes closer, but I'm doubtful that we can really "guarantee" that we will always (in steady state) adhere to the topology specified via the zone configs. The two reasons for that are that atomic replication changes can't move the lease (https://github.com/cockroachdb/cockroach/issues/40333) and the other is my general lack of trust in that the rebalance queue/allocator will generally manage to find correct pairs of stores between which replicas can be moved.

At the time of writing, we'll also definitely violate zone configs when moving replicas off a decommissioning or dead node, because I haven't hooked up atomic replication changes in that path yet. The plan is to do so.

**Comment by danhhz** [comment]
> What I don't understand here is why this talks to the allocator at all; the targets are predetermined by the input.

My guess when I was in there was that it knows which adds and removes it needs to do and it's using the allocator to determine the _order_ it does them.

**Comment by a-robinson** [comment]
> My guess when I was in there was that it knows which adds and removes it needs to do and it's using the allocator to determine the _order_ it does them.

Hi 👋 Yes this is correct. The general idea was that we wanted to do things in alternating steps to avoid adding a ton of replicas at once and causing heavy over-replication (which previously caused problems in early versions of SCATTER). And if we want to do things in alternating steps, there are a number of possible orders in which to do the adds and removes. The decision matters, since some orders are strictly better than others when it comes to considering zone configs and locality diversity. Using the allocator to make these decisions seemed better than writing new, very similar code within `ChangeReplicas` to do so.

**Comment by tbg** [comment]
Thanks Alex!

**Comment by tbg** [comment]
As part of convincing myself that all works as it should, I am planning to use the following SQL query below:

```
------------------------------------------------------------

SELECT
	"uniqueID", count(t) AS repfactor
FROM
	[
		SELECT
			"uniqueID",
			replicas->'node_id' AS n,
			COALESCE(replicas->'type', '0') AS t
		FROM
			system.rangelog,
			jsonb_array_elements(
				info::JSONB->'UpdatedDesc'->'internal_replicas'
			)
				AS replicas
	]
WHERE
	t IN ('0', '2')
GROUP BY
	"uniqueID"
HAVING
	count(t) > 3;
```

This relies on the fact that the `system.rangelog` table is (on short time scales up to weeks) a complete system of record about all of the range descriptors ever written. Range descriptors correspond to replication configurations, so by looking through all of them, we can "prove" that the system never entered any configurations in which there were four voters. That is precisely what the above query (perhaps, inelegantly) does.  Types zero and two correspond to VOTER_FULL and VOTER_INCOMING, respectively. A "regular" addition of a replica onto a triplicated range would add a VOTER_FULL directly, and would be returned by the above query. Similarly, adding a voter via the atomic replication path (but not removing one atomically) would add a VOTER_INCOMING, again causing a row to be emitted. I verified manually in `./cockroach demo --nodes=4` that rebalancing does happen but causes no rows to be returned (i.e. one voter turns into a VOTER_OUTGOING while another one goes from a LEARNER to a VOTER_INCOMING at the same time).

I still need to verify this in larger tests. I know, for example, that at the time of writing RelocateRange (and thus range merges) are going to violate atomic replication changes, and so does decommissioning. The former should go away via #41084 and the latter still needs to be done.

**Comment by tbg** [comment]
My new god query is

```
SELECT
  min("rangeID") AS rangeid,
  min("eventType") AS event,
  count(replica) AS numvoters,
  min("timestamp") AS ts,
  min(info) AS info
FROM
  [
    SELECT
      "uniqueID",
      "timestamp",
      "rangeID",
      "eventType",
      info,
      replica
    FROM
      system.rangelog,
      ROWS FROM (
        jsonb_array_elements(
          info::JSONB->'UpdatedDesc'->'internal_replicas'
        )
      )
        AS replica
    WHERE
      COALESCE(replica->'type', '0') IN ('0', '2')
  ]
WHERE
  "eventType" = 'add'
GROUP BY
  "uniqueID"
HAVING
  count(replica) > 3
ORDER BY
  rangeid, ts ASC;
```

This lists every range that entered a >3x replicated state. I'm running `tpccbench/nodes=9/cpu=4/multi-region` right now to verify that I'm not seeing any of them (other than the ones related to the system ranges).

**Comment by tbg** [comment]
Good and bad news. The good news is that the above query verifies that we're never upreplicating to four nodes in `tpccbench/nodes=9/cpu=4/multi-region`. The bad news is that we're still violating the constraints all the time by putting more than one replica into one AZ.

`tpccbench/nodes=9/cpu=4/multi-region` has nodes 1-3 in region 1, 4-6 in region 2, 7-9 in region 3.

We can check whether we ever put two replicas into region 1 with this query:

```
SELECT
        min("rangeID") AS rangeid,
        min("eventType") AS event,
        count(replica) AS numvoters,
        min("timestamp") AS ts,
        min(info) AS info
FROM
        [
                SELECT
                        "uniqueID",
                        "timestamp",
                        "rangeID",
                        "eventType",
                        info,
                        replica
                FROM
                        system.rangelog,
                        ROWS FROM (
                                jsonb_array_elements(
                                        info::JSONB->'UpdatedDesc'->'internal_replicas'
                                )
                        )
                                AS replica
                WHERE
                        COALESCE(replica->'type', '0') IN ('0', '2')
		AND
			replica->'node_id' IN ('1','2','3')
        ]
WHERE
        "eventType" = 'add'
GROUP BY
        "uniqueID"
HAVING
        count(replica) > 1
ORDER BY
        rangeid, ts ASC;
```

which returns lots of results. Picking a range at random, I see this wonderful behavior:

<img width="856" alt="image" src="https://user-images.githubusercontent.com/5076964/66269066-5a2f9a00-e844-11e9-92ae-ae0ab3559046.png">

Note how *all* of the replicas were put into a single range 💀 

I sort of expected the allocator to do stupid stuff like that, but this is hilariously bad; I hope I broke this just recently. On the "plus" side, the allocator is aware that it's breaking the constraints:

<img width="1323" alt="image" src="https://user-images.githubusercontent.com/5076964/66269083-9f53cc00-e844-11e9-9609-6dd676bc14bc.png">

When the range is split off, with replicas on n3, n4, n8, the constraints should be OK, but then we move the n4 replica to n3, I don't know why, but noting that "constraint check fail".

**Comment by tbg** [comment]
The replicate queue applied to this range at present says

```
2019-10-06 12:25:03 | [n2,s2,r6772/4:/Table/60/2/0{-/1/"AB…}] running replicate.process
-- | --
2019-10-06 12:25:03 | [n2,s2,r6772/4:/Table/60/2/0{-/1/"AB…}] next replica action: consider rebalance
2019-10-06 12:25:03 | [n2,s2,r6772/4:/Table/60/2/0{-/1/"AB…}] no suitable rebalance target
```

This tells me that the allocator doesn't even think there is a problem, which is bad.

**Comment by tbg** [comment]
I found out that this range belongs to the partitioned index `customer_idx`, and so it's intentionally glued to a region via constraints a la

```
ALTER PARTITION p1_0 OF INDEX tpcc.public.customer@customer_idx CONFIGURE ZONE USING
    constraints = '[+zone=us-east1-b]';
```

That makes it even more mysterious why the allocator things the constraints are violated, but at least it's better than I initially thought.

Unfortunately it will be hard to programmatically go through all of the ranges and to decide which ones are supposed to be on which set of replicas.

**Comment by tbg** [comment]
For another bit of confusion, look how many "overreplicated ranges" this thing claims we have:

```
root@localhost:26257/tpcc> select * from system.replication_stats;
  zone_id | subzone_id | report_id | total_ranges | unavailable_ranges | under_replicated_ranges | over_replicated_ranges
+---------+------------+-----------+--------------+--------------------+-------------------------+------------------------+
        0 |          0 |         3 |          198 |                  0 |                       0 |                      0
        1 |          0 |         3 |           18 |                  0 |                       0 |                      0
       15 |          0 |         3 |            1 |                  0 |                       0 |                      0
       16 |          0 |         3 |            1 |                  0 |                       0 |                      0
       17 |          0 |         3 |            2 |                  0 |                       0 |                      0
       22 |          0 |         3 |            1 |                  0 |                       0 |                      0
       53 |          1 |         3 |            5 |                  0 |                       0 |                      5
       53 |          2 |         3 |            6 |                  0 |                       0 |                      6
       53 |          3 |         3 |            4 |                  0 |                       0 |                      4
       54 |          1 |         3 |            1 |                  0 |                       0 |                      0
       54 |          2 |         3 |            1 |                  0 |                       0 |                      0
       54 |          3 |         3 |            1 |                  0 |                       0 |                      0
       55 |          1 |         3 |           21 |                  0 |                       0 |                     21
       55 |          2 |         3 |           20 |                  0 |                       0 |                     20
       55 |          3 |         3 |           22 |                  0 |                       0 |                     22
       56 |          1 |         3 |          434 |                  0 |                       0 |                    434
       56 |          2 |         3 |          424 |                  0 |                       0 |                    424
       56 |          3 |         3 |          428 |                  0 |                       0 |                    428
       56 |          4 |         3 |          243 |                  0 |                       0 |                    243
       56 |          5 |         3 |          251 |                  0 |                       0 |                    251
       56 |          6 |         3 |          251 |                  0 |                       0 |                    251
       57 |          1 |         3 |           89 |                  0 |                       0 |                     89
       57 |          2 |         3 |           88 |                  0 |                       0 |                     88
       57 |          3 |         3 |           87 |                  0 |                       0 |                     87
       57 |          4 |         3 |           48 |                  0 |                       0 |                     48
       57 |          5 |         3 |           47 |                  0 |                       0 |                     47
       57 |          6 |         3 |           47 |                  0 |                       0 |                     47
       57 |          7 |         3 |           42 |                  0 |                       0 |                     42
       57 |          8 |         3 |           40 |                  0 |                       0 |                     40
       57 |          9 |         3 |           41 |                  0 |                       0 |                     41
       58 |          1 |         3 |            2 |                  0 |                       0 |                      2
       58 |          2 |         3 |            2 |                  0 |                       0 |                      2
       58 |          3 |         3 |            2 |                  0 |                       0 |                      2
       59 |          1 |         3 |          709 |                  0 |                       0 |                    709
       59 |          2 |         3 |          705 |                  0 |                       0 |                    705
       59 |          3 |         3 |          692 |                  0 |                       0 |                    692
       60 |          1 |         3 |          379 |                  0 |                       0 |                    379
       60 |          2 |         3 |          377 |                  0 |                       0 |                    377
       60 |          3 |         3 |          377 |                  0 |                       0 |                    377
       60 |          4 |         3 |           56 |                  0 |                       0 |                     56
       60 |          5 |         3 |           53 |                  0 |                       0 |                     53
       60 |          6 |         3 |           56 |                  0 |                       0 |                     56
       61 |          1 |         3 |            1 |                  0 |                       0 |                      1
       61 |          2 |         3 |            1 |                  0 |                       0 |                      1
       61 |          3 |         3 |            1 |                  0 |                       0 |                      1
(45 rows)

Time: 69.719958ms

root@localhost:26257/tpcc>
```

I don't buy that, the `ranges.overreplicated` metric flatlines at zero.

cc @andreimatei / @darinpp (cluster is tobias-1570324285-01-n12cpu4-geo)

**Comment by andreimatei** [comment]
Looking.

**Comment by darinpp** [comment]
https://github.com/cockroachdb/cockroach/issues/41328 shows a case where you have 4 replicas in 19.2. It does seem like a bug however.

**Comment by a-robinson** [comment]
> When the range is split off, with replicas on n3, n4, n8, the constraints should be OK, but then we move the n4 replica to n3, I don't know why, but noting that "constraint check fail".

> ...

> That makes it even more mysterious why the allocator things the constraints are violated, but at least it's better than I initially thought.

In that screenshot, the `details:(constraint check fail)` field is part of the description of the `existing` replica that's being replaced. It's telling you that the reason for removing the replica on s4 and adding a replica on s2 is that s4 violates the constraints of the range.

**Comment by ajwerner** [comment]
@tbg can we close this?
